tests/models/test_hparams.py

```diff
@@ -778,3 +780,42 @@ def test_colliding_hparams(tmpdir):
     trainer = Trainer(default_root_dir=tmpdir, max_epochs=1)
     with pytest.raises(MisconfigurationException, match=r"Error while merging hparams:"):
         trainer.fit(model, datamodule=data)
+
+
+def test_load_checkpoint_update_hparams(tmpdir):
+    """
+    Check that old hparams are updated when new hparams_file is provided
+    """
+
+    class TestModel(BoringModel):
+        def __init__(self, x=None, y=None):
+            super().__init__()
+            self.save_hyperparameters()
+
+    model_checkpoint = ModelCheckpoint(dirpath=tmpdir, filename="{epoch:02d}")
+    trainer = Trainer(
+        max_epochs=1,
+        default_root_dir=tmpdir,
+        limit_train_batches=1,
+        limit_val_batches=1,
+        callbacks=[model_checkpoint],
+        logger=TensorBoardLogger(save_dir=tmpdir),
+    )
+    model = TestModel(5)
+    trainer.fit(model)
+    ckpt_path = os.path.join(tmpdir, "epoch=00.ckpt")
+    hparams_path = os.path.join(tmpdir, "default", "version_0", "hparams.yaml")
+    assert os.path.exists(hparams_path)
+    assert model.hparams.x == 5
+    assert model.hparams.y is None
+    with open(hparams_path, "w") as f:
+        yaml.dump({"y": 10}, f)
```

Comment:
> Depending on the use case, requiring this might suggest that x=5 should be the default value for the model.

Very good point and exactly my concern I tried to express in the linked issue :) 
Printing a warning could mitigate this, however the danger with warnings over errors is that users miss them or outright ignore :)

Reply:
Thanks for the feedback. I'm gonna go ahead and close the issue + PR